src/components/tl/dpu/tl_dpu_context.c

```diff
@@ -0,0 +1,246 @@
+
+            fp = fopen(tl_dpu_config->host_dpu_list, "r");
+            if (fp == NULL) {
+                fprintf(stderr, "Unable to open \"%s\", disabling dpu team\n", tl_dpu_config->host_dpu_list);
+                tl_dpu_config->use_dpu = 0;
+                dpu_found = 0;
+            }
+            else {
+                while (fscanf(fp,"%s", h) != EOF) {
+                    if (strcmp(h, hname) == 0) {
+                        dpu_found = 1;
+                        fscanf(fp, "%s", hname);
+                        fprintf(stderr, "DPU <%s> found!\n", hname);
+                        break;
+                    }
+                    memset(h, 0, 256);
+                }
+            }
+            if (!dpu_found) {
+                tl_dpu_config->use_dpu = 0;
+            }
+        }
+        else {
+            fprintf(stderr, "DPU_ENABLE set, but HOST_LIST not specified. Disabling DPU team!\n");
+            tl_dpu_config->use_dpu = 0;
+        }
+        free(h);
+    }
+    else {
+        goto err;
+    }
+
+    if (!dpu_found) {
+        goto err;
+    }
+
+    fprintf(stdout, "Connecting to %s", hname);
+    sockfd = _server_connect(hname, tl_dpu_config->server_port);
+
+    memset(&ucp_params, 0, sizeof(ucp_params));
+    ucp_params.field_mask      = UCP_PARAM_FIELD_FEATURES |
+                                 UCP_PARAM_FIELD_REQUEST_SIZE |
+                                 UCP_PARAM_FIELD_REQUEST_INIT |
+                                 UCP_PARAM_FIELD_REQUEST_CLEANUP;
+    ucp_params.features        = UCP_FEATURE_TAG |
+                                 UCP_FEATURE_RMA;
+    ucp_params.request_size    = sizeof(tl_dpu_ucp_request_t);
+    ucp_params.request_init    = tl_dpu_ucp_req_init;
+    ucp_params.request_cleanup = tl_dpu_ucp_req_cleanup;
+
+    status = ucp_init(&ucp_params, NULL, &ucp_context);
+    if (status != UCC_OK) {
+        fprintf(stderr, "failed ucp_init(%s)\n", ucc_status_string(status));
+        goto err;
+    }
+
+    memset(&worker_params, 0, sizeof(worker_params));
+    worker_params.field_mask    = UCP_WORKER_PARAM_FIELD_THREAD_MODE;
+    worker_params.thread_mode   = UCS_THREAD_MODE_SINGLE;
+    status = ucp_worker_create(ucp_context, &worker_params, &ucp_worker);
+     if (status != UCC_OK) {
+        fprintf(stderr, "failed ucp_worker_create (%s)\n", ucc_status_string(status));
+        goto err_cleanup_context;
+    }
+
+    worker_attr.field_mask = UCP_WORKER_ATTR_FIELD_ADDRESS |
+                             UCP_WORKER_ATTR_FIELD_ADDRESS_FLAGS;
+    worker_attr.address_flags = UCP_WORKER_ADDRESS_FLAG_NET_ONLY;
+    ucp_worker_query(ucp_worker, &worker_attr);
+
+    ret = send(sockfd, &worker_attr.address_length, sizeof(&worker_attr.address_length), 0);
+    if (ret<0) {
+        fprintf(stderr, "send length failed");
+    }
+    ret = send(sockfd, worker_attr.address, worker_attr.address_length, 0);
+    if (ret<0) {
+        fprintf(stderr, "send address failed");
```

Comment:
i assume if ret < 0 you also need to go to err handling mode, not just print error.